bionic/core/task_execution.py

```diff
@@ -16,46 +32,78 @@
+    milestone: after this, there is no more work to do on this task.
+
+    "Primed": This is a more abstract condition. It guarantees two things:
+
+    a) This task's provenance digest is available, which means any downstream tasks can
+    have their values loaded from the persisted cache. For a task with non-persistable
+    output, its provenance digest depends only on its provenance; it doesn't actually
+    require the task to be computed. However, for a task with persistable output, the
+    provenance digest depends on its actual value, so the task must be computed and its
+    output cached.
+
+    b) There is no additional *persistable* work to do on this task. In other words,
+    if we have any dependent tasks that we plan to run in a separate process, we can
+    go ahead and start them; there may be more work to do on this task, but it will
+    have to be done in that separate process, because its results can't be serialized
+    and transmitted. (On the other hand, if we have a dependent task to run *in this
+    same process*, we'll want to bring this task to the "cached" state instead.) As
+    with (a), for a task with non-persistable output, this milestone is reached as
+    soon as we compute its provenance; for a task with persistable output, it's
+    reached only when the task is computed and its output is cached.
+
+    Thus, the operational definition of "primed" depends on whether the task's output is
+    persistable or not. If so, "primed" is equivalent to "cached"; if not, "primed"
+    represents a preliminary stage of work which happens to be sufficient for many
+    applications.
     """
 
-    def __init__(self, state, is_needed_in_memory):
+    def __init__(self, state):
         self.state = state
         self.stage = EntryStage.NEW
         self.future = None
         self.results_by_dnode = None
 
-        # This is initially set to None as we don't the entire dependency
-        # graph for entries at once. We create it only when it's computed.
-        self._dep_entries = None
+        # This is initially set to None to avoid eagerly recursing through the entire
+        # DAG. We set this once we start processing the entry.
+        self.dep_entries = None
 
-        # If an entry is needed in memory, it cannot be deferred. Only
-        # applicable to non-serializable entries, serializable entries
-        # are never deferred.
-        # NOTE An entry that is not needed in memory by one entry can
-        # be needed in memory by another entry. Depending on our
-        # execution order of such entities, this flag can change from
-        # `False` to `True` in between execution.
-        self.is_needed_in_memory = is_needed_in_memory
+        # These are initially set to False, but can be updated to True when we discover
+        # a requirement from another entry. However, they should never go from True to
+        # False!
+        self.is_requested_to_be_primed = False
```

Comment:
As discussed in a call, we should add a TODO to do the following:
1. Replace these two booleans with an enum.
2. Replace the `EntryBlockage` concept with some kind of `Requirement` object that represents a specific requirement level (using the enum above) on one entry from another.